src/pymoca/tree.py

```diff
@@ -259,6 +290,983 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+        # 4. Lookup classes of extends and ensure it is identical to lookup result from step 3
+        # 5. Check that all children of the current element (including extends) with same name are identical
+        #    (error if not) and only keep one if so (to preserve function argument order)
+        # 6. Components are recursively instantiated
+
+        # 1.1. Partially instantiate the element itself and 1.2 merge modifiers
+        new_class = self._instantiate_partially(
+            orig_class,
+            modification_environment,
+            parent,
+        )
+
+        # 1.3. Redeclare of element itself is done
+        self._apply_class_redeclares(new_class, modification_environment)
+
+        # 2.1 Partially instantiate local classes and symbols
+        if isinstance(orig_class, ast.InstanceClass) and not orig_class.fully_instantiated:
+            from_class = new_class.ast_ref
+        else:
+            from_class = orig_class
+        for name, class_ in from_class.classes.items():
+            instance = self._instantiate_partially(
+                class_,
+                modification_environment,
+                new_class,
+            )
+            new_class.classes[name] = instance
+
+        for name, symbol in from_class.symbols.items():
+            instance = self._instantiate_partially(
+                symbol,
+                modification_environment,
+                new_class,
+            )
+            new_class.symbols[name] = instance
+
+        # 2.2 Copy local contents into the element itself
+        self._copy_class_contents(new_class, copy_extends=True, partially=partially)
+
+        # 3. Instantiate extends and 4. Check extends class lookup
+        for index, extends in enumerate(new_class.extends):
+            extends_instance = self._instantiate_extends(
+                extends, modification_environment, new_class, partially
+            )
+            new_class.extends[index] = extends_instance
+
+        # TODO: Step 5: Check and cull elements with same name in _instantiate_class
+
+        # 6. Recursively instantiate symbols
+        for symbol in new_class.symbols.values():
+            self._instantiate_symbol(symbol, new_class)
+
+        if not partially:
+            new_class.fully_instantiated = True
+
+        return new_class
+
+    def _instantiate_extends(
+        self,
+        extends: ast.ExtendsClause,
+        modification_environment: ast.ClassModification,
+        parent: ast.InstanceClass,
+        partially=True,
+    ) -> ast.InstanceExtends:
+        # 3. For each element in the extends clauses of the current element:
+        #   1. Apply steps 1 and 2 to the element, replacing the extends clause with the extends instance
+        # 4. Lookup classes of extends and ensure it is identical to lookup result from step 3
+
+        extends_class = find_name(extends.component, parent)
```

Comment:
In InstantiationTree, what happens is that the name of the extends clause extends TreeParts.Trunk is looked up in the inherited class itself. That's a bit wasteful (and also confusing). Because we are essentially doing the same lookup twice.

I'd suggest changing this line to 

```Python
extends_class = _find_name(extends.component, parent, current_extends=[extends])
```

Reply:
Fixed in 52d9104